cpp/src/tsne/bh_kernels.cuh

```diff
@@ -701,8 +701,29 @@ __global__ __launch_bounds__(THREADS6, FACTOR6) void IntegrationKernel(
 
     Y1[i] += ux;
     Y2[i] += uy;
+
+    // Sum Ys for centering on the mean later.
+    atomicAdd(&sums[0], Y1[i]);
+    atomicAdd(&sums[1], Y2[i]);
   }
 }
 
+__global__ void mean_center(float *restrict Y1, float *restrict Y2,
```

Comment:
Rather than creating a new kernel for this, we should probably use the `mean_center` primitive in the stats package, even if it means adding a tad extra overhead. This allows us to centralize operations so that we can optimize them in one place instead of having to do it separately for each algorithm.

Reply:
I looked at that, but it would become three kernel invocations because (1) this matrix is oddly shaped (n+1 leading dimension) and (2) this kernel adds noise occasionally.

I did comment on #1389 because that would otherwise be usable here and uses optimized loads/stores